runtime_scan/pkg/orchestrator/scanwatcher/watcher.go

```diff
@@ -0,0 +1,220 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scanwatcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/openclarity/vmclarity/api/models"
+	"github.com/openclarity/vmclarity/runtime_scan/pkg/orchestrator/common"
+	runtimeScanUtils "github.com/openclarity/vmclarity/runtime_scan/pkg/utils"
+	"github.com/openclarity/vmclarity/shared/pkg/backendclient"
+)
+
+const (
+	DefaultPollInterval     = time.Minute
+	DefaultReconcileTimeout = time.Minute
+)
+
+type ScanReconcileEvent struct {
+	ScanID models.ScanID
+}
+
+type (
+	ScanQueue      = common.Queue[ScanReconcileEvent]
+	ScanPoller     = common.Poller[ScanReconcileEvent]
+	ScanReconciler = common.Reconciler[ScanReconcileEvent]
+)
+
+type Config struct {
+	Backend          *backendclient.BackendClient
+	PollPeriod       time.Duration
+	ReconcileTimeout time.Duration
+}
+
+func New(c Config) *Watcher {
+	logger := log.WithFields(log.Fields{"controller": "ScanWatcher"})
+	return &Watcher{
+		logger,
+		c.Backend,
+		c.PollPeriod,
+		c.ReconcileTimeout,
+	}
+}
+
+type Watcher struct {
+	logger           *log.Entry
+	client           *backendclient.BackendClient
+	pollPeriod       time.Duration
+	reconcileTimeout time.Duration
+}
+
+func (w *Watcher) Start(ctx context.Context) {
+	queue := common.NewQueue[ScanReconcileEvent]()
+
+	poller := &ScanPoller{
+		Logger:     w.logger,
+		PollPeriod: w.pollPeriod,
+		Queue:      queue,
+		GetItems:   w.GetAbortedScans,
+	}
+	poller.Start(ctx)
+
+	reconciler := &ScanReconciler{
+		ReconcileTimeout:  w.reconcileTimeout,
+		Queue:             queue,
+		ReconcileFunction: w.Reconcile,
+	}
+	reconciler.Start(ctx)
+}
+
+func (w *Watcher) GetAbortedScans(ctx context.Context) ([]ScanReconcileEvent, error) {
+	scans, err := w.getScansByState(ctx, models.ScanStateAborted)
+	if err != nil || scans.Items == nil || len(*scans.Items) <= 0 {
+		return nil, err
+	}
+
+	count := len(*scans.Items)
+	r := make([]ScanReconcileEvent, count)
+	for i, scan := range *scans.Items {
+		r[i] = ScanReconcileEvent{
+			ScanID: *scan.Id,
+		}
+	}
+
+	return r, nil
+}
+
+func (w *Watcher) Reconcile(ctx context.Context, event ScanReconcileEvent) error {
+	w.logger.Infof("reconciling scan event: %v", event)
+
+	selector := "id,state,stateReason"
+	params := models.GetScansScanIDParams{
+		Select: &selector,
+	}
+
+	scan, err := w.client.GetScan(ctx, event.ScanID, params)
+	if err != nil || scan == nil {
+		return fmt.Errorf("getting scan with id %s failed: %v", event.ScanID, err)
+	}
+
+	state, ok := scan.GetState()
+	if !ok {
+		return fmt.Errorf("cannot determine state of Scan with %s id", event.ScanID)
+	}
+
+	switch state {
+	case models.ScanStateDone, models.ScanStateFailed:
+		w.logger.Infof("reconciling scan event is skipped as Scan is already finished: %v", event)
```

Comment:
Reconciling

Reply:
we want to print that info message every time?

Reply:
I'll set it to debug level, so it won't be log during normal operation.

Reply:
Fixed.